### DISCUSSION TITLE
Temporary Overrides not working properly on new Google cloud build

### DISCUSSION THREAD

**Discussion by idocalm** [NightscoutFoundation/xDrip#2429] [Q&A] [answered]
We've recently migrated to Google Cloud from Heroku. We noticed a strange error related to remote overrides that seems to happen only in the Nightscout version running on Google Cloud's VM. We are using Loop (FreeAPS) on an iPhone 13, the iOS version is 16.0.3. See attached image for more details about the error. Does anyone have any ideas?

![95f6d605-bf1d-43f0-9f28-d3c98070d58c](https://user-images.githubusercontent.com/71544315/200029180-da283d16-033f-47da-a6fc-5e7382ab24b6.jpg)

**Comment by Navid200** [comment]


**Reply by idocalm** [reply]
Done. Sorry.

**Comment by Navid200** [comment, marked as answer]
Have you set up the tokens correctly in the app that implements remote overrides?

**Reply by idocalm** [reply]
I'm not sure how would one do that, but we followed this official guide: https://loopkit.github.io/loopdocs/nightscout/remote-overrides/#step-1-update-the-loopers-iphone-settings
The tutorial instructed us to add new config vars: 
- LOOP_APNS_KEY
- LOOP_APNS_KEY_ID
- LOOP_DEVELOPER_TEAM_ID
or (these are made-up values from the tutorial and will not work) : 
![Screenshot 2022-11-05 103716](https://user-images.githubusercontent.com/71544315/200112207-524c2ad3-5b8e-43a1-b3b2-a9b5daff70bf.png)

*I have thought of something, that might cause the problem.*
The Google Cloud VM will only implement the config var if it's a one-line, but as you can see in the image, the LOOP_APNS_KEY will communicate well with apple if it's exactly the same, including the line breaks and spaces.

As @Navid200 requested, here are the current config vars:
(take a look at how I manipulated the LOOP_APNS_KEY using \n, maybe that's wrong?)

```
export AUTH_DEFAULT_ROLES="admin"
export ALARM_HIGH="false"
export ALARM_LOW="false"
export ALARM_TIMEAGO_URGENT="off"
export ALARM_TIMEAGO_URGENT_MINS="30"
export ALARM_TIMEAGO_WARN="off"
export ALARM_TIMEAGO_WARN_MINS="15"
export ALARM_TYPES="simple"
export ALARM_URGENT_HIGH="off"
export ALARM_URGENT_LOW="off"
export API_SECRET="MY_API_SECRET"
export BASAL_RENDER="default"
export BG_HIGH="200"
export BG_LOW="60"
export BG_TARGET_BOTTOM="90"
export BG_TARGET_TOP="180"
export BOLUS_RENDER_OVER="1"
export BRIDGE_INTERVAL="150000"
export DEVICESTATUS_ADVANCED="true"
export DISPLAY_UNITS="mg/dl"
export ENABLE='careportal basal dbsize rawbg iob maker cob bwp cage iage sage boluscalc pushover treatmentnotify loop pump profile food openaps bage alexa override speech cors bridge'
export LOOP_APNS_KEY="MYKEY"
export LOOP_APNS_KEY_ID="JDJC9QTLFD"
export LOOP_DEVELOPER_TEAM_ID="D8UBP4824W"
export LOOP_ENABLE_ALERTS="true"
export LOOP_URGENT="60"
export LOOP_WARN="20"
export NIGHT_MODE="off"
export PAPERTRAIL_API_TOKEN="IZID6K9VfBVAya32y64"
export PUMP_ENABLE_ALERTS="true"
export PUMP_FIELDS='battery reservoir clock status'
export PUMP_RETRO_FIELDS='battery reservoir clock status'
export PUMP_URGENT_BATT_U="30"
export PUMP_URGENT_BATT_V="1.25"
export PUMP_URGENT_CLOCK="30"
export PUMP_URGENT_RES="10"
export SHOW_FORECAST="loop"
export SHOW_PLUGINS='loop pump cob iob sage cage careportal basal override dbsize'
export SHOW_RAWBG="never"
export THEME="colors"
export TIME_FORMAT="24"
```

**Reply by Navid200** [reply]
Please open your variables on Heroku.  Click on the pencil symbol for the LOOP-APN-KEY variable.  Highlight the entire content and copy it to your clipboard.
Now, close the variable without changing anything.

Now, open a temporary textfile and paste in in.
What does it look like if you extend the window containing the textfile horizontally?

Does it only contain three lines with the second line being very long?
Regardless, edit the Nightscout variables on Google Cloud and just paste the content you copied from Heroku, exactly as it is, to it inside the quotation marks for the particular variable.
So, there should be no added \n in it.  If there is only a single line in the textfile you created, there has to be a single line in the nsconfig file.
If there are three lines, pasting it will create three lines.  In that case, the quotation mark opens on the first line and closes on the third.

**Reply by idocalm** [reply]
Thank you so much Navid, pasting the config var exactly like in Heroku actually fixed the problem. ❤️

**Reply by Navid200** [reply]
Great
Please kindly, if you haven't already, mention the same on facebook as there is no association between the two forums.
That way if someone with the same problem finds your thread on facebook, will see the answer as well.

**Reply by austinbaughman** [reply]
Hello, it seems we are having the same issue. I'm coming from heroku (can't access it now) - to google cloud using NS on freeAPS. We do not have the override option anymore. This was very helpful for us as we help my daughters teacher manager her levels while she is act school. Would anyone like to help us get the override option working on our nightscout site? I'm on facebook under austinbaughman - or you can email me here: austin.baughman@tctexas.org | My daughter is 6 years old. We were not aware of the heroku change so the last few days have been a little troubling for u. we do have NS up and running--just no override and one other small issue we're troubleshooting. thanks in advance

**Reply by austinbaughman** [reply]
> Thank you so much Navid, pasting the config var exactly like in Heroku actually fixed the problem. ❤️

Can you share the config var with me so i can try to do this? I'm not sure how to but i'm going to try!

**Reply by idocalm** [reply]
Of course. @tzachi-dar I think Tzachi might be able to help you more since he was dealing with pretty much the same problem. If you have set up the APNS keys as required in the remote override tutorial: 
https://loopkit.github.io/loopdocs/nightscout/remote-overrides/
*I think you might just need to wait a few hours...*

My config vars are listed in this discussion - above this reply
https://github.com/NightscoutFoundation/xDrip/discussions/2429#discussioncomment-4062826

~~Also, check if you're using Nightscout 14.2.6, it might have something to do with this issue (not confirmed), try changing to 14.2.5~~

**Reply by tzachi-dar** [reply]
If you had temporary overrides working on heroku, than you just need to copy the vars from heroku to NS.
The easiest way to do it (If you don't have 2FA) is to run the script /xDrip/scripts/GetHerokuVars.sh
Alternatively, you can edit the file  /etc/nsconfig directly and put the parameters there. You need to make sure that the line breaks are kept as is.
Using the web tool (to copy the variables) is currently not possible, because it does not support multiline data.
It will be supported in the next version (probably 2/3 days). If you know how, you can use it from https://github.com/jamorham/nightscout-vps/pull/28 and use it.

**Reply by tzachi-dar** [reply]
one more thing, don't worry about the 14.2.6 version, once you set everything correctly, and wait 2-3 hours all will work on that version as well.

**Reply by austinbaughman** [reply]
> If you had temporary overrides working on heroku, than you just need to copy the vars from heroku to NS. The easiest way to do it (If you don't have 2FA) is to run the script /xDrip/scripts/GetHerokuVars.sh Alternatively, you can edit the file /etc/nsconfig directly and put the parameters there. You need to make sure that the line breaks are kept as is. Using the web tool (to copy the variables) is currently not possible, because it does not support multiline data. It will be supported in the next version (probably 2/3 days). If you know how, you can use it from [jamorham/nightscout-vps#28](https://github.com/jamorham/nightscout-vps/pull/28) and use it.

Thank You.. I ran the script and got this
"Logged in as austinbaughman@gmail.com
austinbaughman
new vars are in the file heroku_vars.txt.
Please look at it, and if all is ok, try copying it to /etc/nsconfig"
Does this mean it copied over correctly?

**Reply by blameitonben** [reply]
So I've finally got to a point where my GC is seeing the overrides, but now I'm receiving the error "Failed to Generate Token" My Heroku site was deleted long ago, I've created a new key from apple, but I'm guessing the copy and past isn't formatting correctly from the original text editor? This has been such a frustrating issue trying to resolve since there's so little info on the Google cloud side of this, if it's a formatting issue, can someone please give good detail on how it should be formatted from the downloaded text file from apple?

**Comment by idocalm** [comment]
@Navid200 Just talked to someone trying to install Nightscout on Google Cloud, he installed version 14.2.6 while I installed version 14.2.5. We have the same exact config vars, but his Nightscout website appears to not even show Temporary Overrides as an option in the treatments tab and mine is working perfectly. Might be worth checking.

**Reply by austinbaughman** [reply]
How would i go back to 14.2.5? Because we are not seeing override either on 14.2.6

**Reply by Navid200** [reply]
You can go to a different version by creating a GitHub fork of the version.  Then, you can use the Customize Nightscout option on the menu to install from your fork.
Customizing Nightscout can take up to 30 minutes.  If there is an interruption, you will need to repeat.

I don't recommend that unless it is approved that's the only way to resolve the issue you are experiencing.

**Comment by tzachi-dar** [comment]
I have tried the suggestion suggested by @Navid200, and on a first look it did not work. (we have been using 14.2.5) the nightscout site did not show temporary overrides as well. We have left it running for 2-3 hours and everything works. So, I'm not sure what happened, but sometimes patience helps.

**Reply by mtndev** [reply]
Thanks for posting this.. i'd about given up hope but since I had a heads  up from you that there could be a delay, I waited and then after about 1-2 hours, it just showed up when I refreshed the page.
